packages/firestore/src/local/tab_notification_channel.ts

```diff
@@ -102,6 +171,81 @@ export class LocalStorageNotificationChannel implements TabNotificationChannel {
     this.persistState();
   }
 
+  addMutation(batchId: BatchId): void {
+    this.pendingBatchIds[batchId] = this.instanceId;
+  }
+
+  rejectMutation(batchId: BatchId, error: FirestoreError): void {
+    delete batchId[batchId];
+    this.rejectedBatchIds[batchId] = { date: new Date(), error };
+  }
+
+  acknowledgeMutation(batchId: BatchId): void {
+    this.acknowledgedBatchIds[batchId] = new Date();
+  }
+
+  addQuery(targetId: TargetId): void {
+    this.activeTargetIds[targetId] = this.instanceId;
+  }
+
+  removeQuery(targetId: TargetId): void {
+    delete this.activeTargetIds[targetId];
+  }
+
+  rejectQuery(targetId: TargetId, error: FirestoreError): void {
+    this.rejectedTargetIds[targetId] = { date: new Date(), error };
+  }
+
+  updateQuery(updatedTargetIds: TargetId[]): void {
+    for (const targetId of updatedTargetIds) {
+      this.updatedTargetIds[targetId] = new Date();
+    }
+  }
+
+  private onUpdate(key: string, value: string) {
```

Comment:
Your pseudocode needs more comments and less code. :-) Who calls this? when?

Reply:
All these methods are supposed to call this.persistState(). Comment added.

Reply:
This is the event listener hook called when a key in LocalStorage changes.